### DISCUSSION TITLE
RoadRunner silently crashes leaving defunct process

### DISCUSSION THREAD

**Discussion by egonbraun** [roadrunner-server/roadrunner#1218] [Q&A]
Hi, I have a backend application written in `PHP 8.1` and `Laravel v9.19.0`. We were using NginX+FPM and now we are trying to migrate to NginX with Octane+RoadRunner but we are running into an issue that I need your help with.

So, we are using `RoadRunner 2.10.6` and our controller function is running fine, however when the controller function returns the `Illuminate\Http\Response` object RoadRunner crashes and NginX returns a 500 error. No logs, no stack traces, nothing. I just see the 500 error in access logs.

I also noticed that RoadRunner processes are becoming defunct (zombie). When I run `ps auxww` I see dozens of processes like this one:

```bash
root     12701  0.0  0.0      0     0 ?        Z    14:47   0:00 [rr] <defunct>
```

This is how the RoadRunner process is being executed by Octane:

```bash
/usr/bin/rr -c /app/.rr.yaml -o version=2.7 -o http.address=127.0.0.1:8000 -o http.access_logs=true -o http.pool.debug=true -o http.pool.max_jobs=500 -o http.pool.num_workers=0 -o http.pool.supervisor.exec_ttl=300s -o logs.mode=development -o logs.level=info -o logs.output=stderr -o logs.err_output=stderr -o logs.encoding=console -o rpc.listen=tcp://:6001 -o server.command=/usr/bin/php8.1 /app/vendor/bin/roadrunner-worker -o server.user=www-data -o server.group=www-data -o server.relay=pipes serve
```

I could use some ideas on how to further troubleshoot this issue.

Thank you very much.

**Comment by rustatian** [comment]
Hey @egonbraun 👋🏻 

`-o http.pool.debug=true` this option should be set to `true` only for the development. For some reason, processes exited before the `wait4` syscall and became zombies in that case.

**Comment by rustatian** [comment]
Could you please remove that option?

**Comment by egonbraun** [comment]
Sure, doing it now :)

**Comment by egonbraun** [comment]
I believe that did it! Thanks, it is still failing but now I see why it is failing and I gotta do some digging on my end to find the root cause. Thank you very much @rustatian!

**Reply by rustatian** [reply]
My pleasure 🤝 
Keep in mind, that these options might be used for debugging (or local development), however, they should be updated for the production:

           DEV                
`-o logs.mode=development`      
`-o http.access_logs=true`      
`-o logs.level=info`            

          PROD
`-o logs.mode=production`
`-o http.pool.debug=true` -> removed or set to false.
`-o http.access_logs=false` (however, might be turned on to debug) 
`-o logs.level=error`

For the logs, you might use channels to precisely configure the logger. For example: `-o logs.channels.http.level=error` will set the only HTTP log level to `error` (https://github.com/roadrunner-server/roadrunner/blob/master/.rr.yaml#L146)

**Reply by egonbraun** [reply]
Maybe I have celebrated too soon. =P

Here is how RoadRunner is now being called by Octane:

```
/usr/bin/rr /usr/bin/rr \
  -c /app/.rr.yaml \
  -o version=2.7 \
  -o http.address=127.0.0.1:8000 \
  -o http.access_logs=true \
  -o http.pool.max_jobs=500 \
  -o http.pool.num_workers=0 \
  -o http.pool.supervisor.exec_ttl=300s \
  -o logs.mode=production \
  -o logs.level=debug \
  -o logs.output=stderr \
  -o logs.err_output=stderr \
  -o logs.encoding=console \
  -o rpc.listen=tcp://:6001 \
  -o server.command=/usr/bin/php8.1 /app/vendor/bin/roadrunner-worker \
  -o server.user=www-data \
  -o server.group=www-data \
  -o server.relay=pipes \
  serve
```

Note that the `/app/.rr.yaml` file is empty.

Here are the logs I am seeing now after implementing the changes you mentioned. These logs appear after I call my healthcheck endpoint.

```
nginx access_log 169.254.170.1 "GET /healthcheck-all HTTP/1.1" 500 "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.5060.114 Safari/537.36" 0 7.812

ERROR  1.6582237034479065e+09	info	http	http access log	{"read_bytes": 0, "write_bytes": 0, "status": 500, "method": "GET", "URI": "/healthcheck-all", "remote_address": "127.0.0.1:47064", "query": "", "content_len": 0, "host": "localhost:5000", "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.5060.114 Safari/537.36", "referer": "", "time_local": "19/Jul/22:09:41:43 +0000", "request_time": 1658223703.4478815, "start": 1658223695.658189, "elapsed": 7.789833253}

ERROR  1.6582237034885225e+09	debug	server	worker stopped	{"internal_event_name": "EventWorkerWaitExit", "error": "signal: killed; process_wait: signal: killed", "errorCauses": [{"error": "signal: killed"}, {"error": "process_wait: signal: killed"}]}

ERROR  1.6582237040637803e+09	debug	server	worker is allocated	{"pid": 2452, "internal_event_name": "EventWorkerConstruct"}
```

By the way, not sure if it matters but this is running inside a Docker container.

```
$ ps auxww | grep -i rr
root       217  0.5  0.7 904396 60500 ?        Sl   09:36   0:05 /usr/bin/qemu-x86_64 /usr/bin/rr /usr/bin/rr -c /app/.rr.yaml -o version=2.7 -o http.address=127.0.0.1:8000 -o http.access_logs=true -o http.pool.max_jobs=500 -o http.pool.num_workers=0 -o http.pool.supervisor.exec_ttl=300s -o logs.mode=production -o logs.level=debug -o logs.output=stderr -o logs.err_output=stderr -o logs.encoding=console -o rpc.listen=tcp://:6001 -o server.command=/usr/bin/php8.1 /app/vendor/bin/roadrunner-worker -o server.user=www-data -o server.group=www-data -o server.relay=pipes serve
root       589  0.0  0.0      0     0 ?        Z    09:38   0:00 [rr] <defunct>
root      1691  0.0  0.0      0     0 ?        Z    09:40   0:00 [rr] <defunct>
root      1750  0.0  0.0      0     0 ?        Z    09:40   0:00 [rr] <defunct>
```

**Reply by rustatian** [reply]
Yeah, ok :) If that possible to reproduce this, could you please create a repository with the sample and steps? It might be `octane` killing workers instead of stopping them...

**Reply by rustatian** [reply]
And create a bug for this.

**Reply by rustatian** [reply]
I guess that on the `500` status from the worker, the octane kills it because RR doesn't interrupt the worker if it returns any status code, but without error.

**Reply by rustatian** [reply]
And this is not an `ERROR` in the RR terms. I'm not a PHP dev, but could you please check what is octane doing on the `500` returned from the worker?

**Reply by egonbraun** [reply]
Sure, I will do that. I get the feeling that it also might be something in our codebase this is why I first came to the Q&A section. I tried reproducing the issue using a fresh new Laravel+Octane installation but I couldn't.

I will see if I can create a project so you guys can reproduce it over there.

**Reply by egonbraun** [reply]
> And this is not an `ERROR` in the RR terms. I'm not a PHP dev, but could you please check what is octane doing on the `500` returned from the worker?

Yep, I am also not a PHP Dev so I am trying to get someone to help me run a debug session to see where exactly this fails.

Thanks for your help @rustatian, I appreciate your prompt response!
